<a href="https://colab.research.google.com/github/cotozelo/DataScience_Python_-_Data_ICMC/blob/main/Notebooks/Introdu%C3%A7%C3%A3o_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pandas

Esse é um notebook, que tem a finalidade de fazer um introdução das principais usos da biblioteca pandas (https://pandas.pydata.org/docs/).

Os exemplos desse notebook foram seguidos do curto de "Introdução à Data Science com Python" do canal de Youtube Data ICMC (https://www.youtube.com/playlist?list=PLFE-LjWAAP9SfEuLXf3qrpw4szKWjlYq9).

Na sequencia vermos os exemplos...

## Agenda

* Imports
* Ler arquivo .csv



## imports

In [4]:
import numpy as np
import pandas as pd # é o import para usar o pandas, comumente renomeamos como pd

## Ler arquivo .csv

Antes de rodar as células a baixo é preciso salvar o arquivo na pasta do colab.

In [5]:
# Efetuando a leitura do arquivo .csv e salvamos na variavl df
df = pd.read_csv('titanic.csv')

In [6]:
# Verificando o tipo de da variavel df
type(df)

pandas.core.frame.DataFrame

##O DataFrame

###.head() .tail()
Para ver uma parte do dataframe podemos usar o método .head(). Esse método mostra as primeiras linhas do df, por padrão são as 5 primeiras linhas, mas podemos estilar a quanditdade desejada. 
O método .tail() funciona semelhante ao .head(), mas nesse caso mostra as últimas linhas, e seu padrão tb são as 5 últimas.

In [7]:
#mosntrando as 5 primeiras
df.head()

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

In [8]:
#Podemos estipular a quantidade de linhas
df.head(10)

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S
5            6         0       3  ...   8.4583   NaN         Q
6            7         0       1  ...  51.8625   E46         S
7            8         0       3  ...  21.0750   NaN         S
8            9         1       3  ...  11.1333   NaN         S
9           10         1       2  ...  30.0708   NaN         C

[10 rows x 12 columns]

###shape
Para saber a estrutra dimensional de um df, podemos verificar o atributo .shape, ele retorna uma tupla, onde a primeira posição é a linha e a segunda posição a colunas.
(linha, coluna)
(213, 21)

In [9]:
df.shape

(891, 12)

###.columns & .dtypes
O atributo columns retorna todas as colunas do dataframe. Também por ele podemos renomear as colunas.

O atributo .dtypes retorna qual o tipo para cada uma das colunas. Tipo dos dados salvos nela.

In [10]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [11]:
colunas_original = df.columns
colunas_original

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [12]:
df.columns = ['PassengerId_new', 'Survived_new', 'Pclass_new', 'Name_new', 'Sex_new', 'Age_new', 'SibSp_new',
       'Parch_new', 'Ticket_new', 'Fare_new', 'Cabin_new', 'Embarked_new']
df.columns

Index(['PassengerId_new', 'Survived_new', 'Pclass_new', 'Name_new', 'Sex_new',
       'Age_new', 'SibSp_new', 'Parch_new', 'Ticket_new', 'Fare_new',
       'Cabin_new', 'Embarked_new'],
      dtype='object')

In [13]:
#voltando as colunas originais
df.columns = colunas_original
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [14]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

###.descrebi()
Esse método retorna varias métricas estatisticas sobre cada uma das colunas.
As métricas são: quantidade, média, desvio padrão, melhor valor, maior valor, 1º quatil, 2º quatil e 3º quatil. Essas métrica são refletidas somente em columas numéricas, e note que em algumas colunas o count tem valores diferente, isso por que o count soma somente as linhas cujos valores são não nulos.

In [15]:
df.describe()

PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

In [16]:
#diferença entre média e mediana
numeros = [8,5,3,4,9,10,11]
print(f'num1: {numeros}')
print(f'Média   num1: {np.mean(numeros)}')
print(f'Mediana num1: {np.median(numeros)}')
print()
numeros = [8,5,3,4,9,10,1100]
print(f'num2: {numeros}')
print(f'Média   num2: {np.mean(numeros)}')
print(f'Mediana num2: {np.median(numeros)}')

print("Note que a mediana consegue ser mais estável caso haja um valor muito discrepante.")

num1: [8, 5, 3, 4, 9, 10, 11]
Média   num1: 7.142857142857143
Mediana num1: 8.0

num2: [8, 5, 3, 4, 9, 10, 1100]
Média   num2: 162.71428571428572
Mediana num2: 8.0
Note que a mediana consegue ser mais estável caso haja um valor muito discrepante.


Mais referência sobre métricas estatísticas segue o link: http://portalaction.com.br/estatistica-basica/estatisticas-descritivas

##Funções de agregação
A operação das funções é feita individualmente por cada coluna.

* sum
* min
* mean

In [17]:
df.sum()

PassengerId                                               397386
Survived                                                     342
Pclass                                                      2057
Name           Braund, Mr. Owen HarrisCumings, Mrs. John Brad...
Sex            malefemalefemalefemalemalemalemalemalefemalefe...
Age                                                      21205.2
SibSp                                                        466
Parch                                                        340
Ticket         A/5 21171PC 17599STON/O2. 31012821138033734503...
Fare                                                     28693.9
dtype: object

In [18]:
df.mean()

PassengerId    446.000000
Survived         0.383838
Pclass           2.308642
Age             29.699118
SibSp            0.523008
Parch            0.381594
Fare            32.204208
dtype: float64

In [19]:
df.min()

PassengerId                      1
Survived                         0
Pclass                           1
Name           Abbing, Mr. Anthony
Sex                         female
Age                           0.42
SibSp                            0
Parch                            0
Ticket                      110152
Fare                             0
dtype: object

##Verificação de valores nulos
Uma operação que sempre devemos fazer nos dados, é a verificação de valores nulos, para isso podemos usar o método .isna(), segue exemplos.

In [20]:
df.isna()

PassengerId  Survived  Pclass   Name  ...  Ticket   Fare  Cabin  Embarked
0          False     False   False  False  ...   False  False   True     False
1          False     False   False  False  ...   False  False  False     False
2          False     False   False  False  ...   False  False   True     False
3          False     False   False  False  ...   False  False  False     False
4          False     False   False  False  ...   False  False   True     False
..           ...       ...     ...    ...  ...     ...    ...    ...       ...
886        False     False   False  False  ...   False  False   True     False
887        False     False   False  False  ...   False  False  False     False
888        False     False   False  False  ...   False  False   True     False
889        False     False   False  False  ...   False  False  False     False
890        False     False   False  False  ...   False  False   True     False

[891 rows x 12 columns]

O retorno do .isna() é um bool que informa True ou False para todas as posições, com isso sabemos quais posições são nulas ou não.

Também podemos contar as posições nulas, por colunas, segue:

In [21]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Podemos constatar agora que as colunas 'Age', 'Cabin' e 'Embarked', contém valores nulos de 177, 687 e 2 registros respectivamente.

##Indexação
Mecanismos de acessar uma posição no dataframe. Existe varios modos de acessar posições.

Mais referências sobre indexação pode ser obtida no link: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html


In [22]:
#Criando um df
df_toy = pd.DataFrame(data=np.random.randint(20, size=(8, 4)),
                      index=['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'],
                      columns=['Col 1', 'Col 2', 'Col 3', 'Col 4'])

df_toy

Col 1  Col 2  Col 3  Col 4
a     14      7      0      4
b      1      4     15      8
c      3      3     19     15
d      8      4      8     15
e     12      6      3     17
f      8     11      6     14
g      8     18     17     19
h      6      3      6      5

###Obter coluna

In [23]:
col_1 = df_toy['Col 1']
col_1

a    14
b     1
c     3
d     8
e    12
f     8
g     8
h     6
Name: Col 1, dtype: int64

Quando obtemos uma única columa o tipo de retorno é um Series, note a seguir:

In [24]:
type(col_1)

pandas.core.series.Series

In [25]:
col_1_2 = df_toy[['Col 1', 'Col 2']]
col_1_2

Col 1  Col 2
a     14      7
b      1      4
c      3      3
d      8      4
e     12      6
f      8     11
g      8     18
h      6      3

Agora o retorno (col_1_2) é um datafame, note a seguir:

In [26]:
type(col_1_2)

pandas.core.frame.DataFrame

###Obtendo linhas

In [27]:
df_toy[2:4]

Col 1  Col 2  Col 3  Col 4
c      3      3     19     15
d      8      4      8     15

O retorno são somentes as linhas selecionados, note que o 2 é inclusivo (intervalo fechado) e o 4 é exclusivo (intervalo aberto), e o retorno é também um dataframe.

In [28]:
type(df_toy[2:4])

pandas.core.frame.DataFrame

In [29]:
df_toy[[True, True, False, False, True, False, False, True]]

Col 1  Col 2  Col 3  Col 4
a     14      7      0      4
b      1      4     15      8
e     12      6      3     17
h      6      3      6      5

Podemos passar um lista com bool, informando para cada linha se ela deve (True) ou não (False) ser exibida, note que quantidade de elementos da lista de ser igual a quantidade de linhas do df.

In [30]:
maior_10 = df_toy['Col 1'] > 10
maior_10

a     True
b    False
c    False
d    False
e     True
f    False
g    False
h    False
Name: Col 1, dtype: bool

In [31]:
df_toy[maior_10]

Col 1  Col 2  Col 3  Col 4
a     14      7      0      4
e     12      6      3     17

Possibilita a busca dentro do df.

###.loc (indexa por label)


In [32]:
df_toy.loc['a']

Col 1    14
Col 2     7
Col 3     0
Col 4     4
Name: a, dtype: int64

In [33]:
df_toy.loc['a',:]

Col 1    14
Col 2     7
Col 3     0
Col 4     4
Name: a, dtype: int64

Não diferença entre fazermos df_toy.loc['a'] ou df_toy.loc['a',:]. O segundo abre a possibilidade de filtramos algumas colunas, ao passo que o primeiro retorna sempre todas as colunas.

In [34]:
df_toy.loc[['a', 'b'], ['Col 1', 'Col 2', 'Col 3']]

Col 1  Col 2  Col 3
a     14      7      0
b      1      4     15

###.iloc (indexa por posição, indice)


In [35]:
#Seleciona a primeira linha e todas as colunas
df_toy.iloc[0, :]

Col 1    14
Col 2     7
Col 3     0
Col 4     4
Name: a, dtype: int64

In [36]:
#Seleciona a segunda coluna todas as linhas
df_toy.iloc[:, 1]

a     7
b     4
c     3
d     4
e     6
f    11
g    18
h     3
Name: Col 2, dtype: int64

In [37]:
#Seleção por lista de linhas e colunas
df_toy.iloc[[2, 3, 4], [0, 2]]

Col 1  Col 3
c      3     19
d      8      8
e     12      3

In [38]:
#Seleção por lista bool
df_toy.iloc[:, [True, False, False, True]]

Col 1  Col 4
a     14      4
b      1      8
c      3     15
d      8     15
e     12     17
f      8     14
g      8     19
h      6      5

###Adicionando

####Colunas

In [41]:
#crinado (ou subistituindo) uma coluna com valores aleatórios 
df.loc[:, 'Coluna Aleatoria'] = np.random.randint(100, size=df.shape[0])
df.head()

PassengerId  Survived  Pclass  ... Cabin Embarked  Coluna Aleatoria
0            1         0       3  ...   NaN        S                34
1            2         1       1  ...   C85        C                13
2            3         1       3  ...   NaN        S                28
3            4         1       1  ...  C123        S                 7
4            5         0       3  ...   NaN        S                94

[5 rows x 13 columns]

In [44]:
#funciona igual ao comando com .loc, mas a recomendação é usar .loc
df['Coluna Aleatoria2'] = np.random.randint(100, size=df.shape[0])
df.head()

PassengerId  Survived  Pclass  ... Embarked Coluna Aleatoria  Coluna Aleatoria2
0            1         0       3  ...        S               34                 66
1            2         1       1  ...        C               13                 70
2            3         1       3  ...        S               28                 57
3            4         1       1  ...        S                7                 30
4            5         0       3  ...        S               94                 51

[5 rows x 14 columns]

####Linhas

In [65]:
df.loc[df.shape[0],:] = np.random.randint(100, size=len(df.columns))
df.tail()

PassengerId  Survived  Pclass  ... Cabin Embarked  Coluna Aleatoria
882        883.0       0.0     3.0  ...   NaN        S              96.0
883        884.0       0.0     2.0  ...   NaN        S              29.0
884        885.0       0.0     3.0  ...   NaN        S              65.0
885        886.0       0.0     3.0  ...   NaN        Q              80.0
886         79.0      23.0    86.0  ...    67       69              77.0

[5 rows x 13 columns]

###Remoção

####Linhas

No .drop() vc informa a linha que deseja remover, mas ele funciona como um filtro, ou seja, ele não altera o df em questão, ele só devolve um df a linha. Para aplicar no próprio df precisa setar o parâmetro inplace.

In [45]:
df.drop(0)

PassengerId  Survived  ...  Coluna Aleatoria Coluna Aleatoria2
1              2         1  ...                13                70
2              3         1  ...                28                57
3              4         1  ...                 7                30
4              5         0  ...                94                51
5              6         0  ...                57                35
..           ...       ...  ...               ...               ...
886          887         0  ...                35                26
887          888         1  ...                68                65
888          889         0  ...                63                64
889          890         1  ...                33                25
890          891         0  ...                83                31

[890 rows x 14 columns]

In [46]:
df.head()

PassengerId  Survived  Pclass  ... Embarked Coluna Aleatoria  Coluna Aleatoria2
0            1         0       3  ...        S               34                 66
1            2         1       1  ...        C               13                 70
2            3         1       3  ...        S               28                 57
3            4         1       1  ...        S                7                 30
4            5         0       3  ...        S               94                 51

[5 rows x 14 columns]

In [48]:
df2 = df.drop(0)
df2.head()

PassengerId  Survived  Pclass  ... Embarked Coluna Aleatoria  Coluna Aleatoria2
1            2         1       1  ...        C               13                 70
2            3         1       3  ...        S               28                 57
3            4         1       1  ...        S                7                 30
4            5         0       3  ...        S               94                 51
5            6         0       3  ...        Q               57                 35

[5 rows x 14 columns]

In [49]:
df.drop(0, inplace=True)
df.head()

PassengerId  Survived  Pclass  ... Embarked Coluna Aleatoria  Coluna Aleatoria2
1            2         1       1  ...        C               13                 70
2            3         1       3  ...        S               28                 57
3            4         1       1  ...        S                7                 30
4            5         0       3  ...        S               94                 51
5            6         0       3  ...        Q               57                 35

[5 rows x 14 columns]

####Colunas

Para remover colunas também usamos o drop, mas infornado explicitamente qual coluna desejamos remover.
A regra do inplace se mantêm.

In [50]:
df.drop(columns=['Coluna Aleatoria2'], inplace=True)
df.head()

PassengerId  Survived  Pclass  ... Cabin Embarked  Coluna Aleatoria
1            2         1       1  ...   C85        C                13
2            3         1       3  ...   NaN        S                28
3            4         1       1  ...  C123        S                 7
4            5         0       3  ...   NaN        S                94
5            6         0       3  ...   NaN        Q                57

[5 rows x 13 columns]

##Tratamento de nulos

In [70]:
nulos = df.isna().sum()
nulos

PassengerId           0
Survived              0
Pclass                0
Name                  0
Sex                   0
Age                 176
SibSp                 0
Parch                 0
Ticket                0
Fare                  0
Cabin               682
Embarked              2
Coluna Aleatoria      0
dtype: int64

In [72]:
#Percentual de valores faltantes, valores em percentuais
100 * (nulos / df.shape[0])

PassengerId          0.000000
Survived             0.000000
Pclass               0.000000
Name                 0.000000
Sex                  0.000000
Age                 19.864560
SibSp                0.000000
Parch                0.000000
Ticket               0.000000
Fare                 0.000000
Cabin               76.975169
Embarked             0.225734
Coluna Aleatoria     0.000000
dtype: float64

####Subistituição pela média

Assim todos os valores NA's serão substituidos pela média dos valores não NA.

In [77]:
idade_media = df['Age'].mean()
df['Age'].fillna(idade_media, inplace=True)
df.head()

PassengerId  Survived  Pclass  ... Cabin Embarked  Coluna Aleatoria
1          2.0       1.0     1.0  ...   C85        C              13.0
2          3.0       1.0     3.0  ...   NaN        S              28.0
3          4.0       1.0     1.0  ...  C123        S               7.0
4          5.0       0.0     3.0  ...   NaN        S              94.0
5          6.0       0.0     3.0  ...   NaN        Q              57.0

[5 rows x 13 columns]

In [79]:
df['Age'].isna().sum()

0

####Remoção da coluna
Para o caso de Cabin que tem mais de 76% de valores faltantes, seria interessante remover toda a coluna.

In [80]:
#axis=1 é para retirar as colunas
#thresh=600 retirar quanto tiver mais que 600 ocorrências
df.dropna(axis=1, thresh=600, inplace=True)
df.isna().sum()

PassengerId         0
Survived            0
Pclass              0
Name                0
Sex                 0
Age                 0
SibSp               0
Parch               0
Ticket              0
Fare                0
Embarked            2
Coluna Aleatoria    0
dtype: int64

####Retirar linhas
No caso de Embarked que só tem 0.22% de valores faltantes, faz sentido retirar esse registros, pois isso não impacta na quantidade da base.

In [82]:
#axis=0 é para retirar as linhas
#subset=['Embarked'] olhar somente nessa colunas Embarked e retirar os NA dela
df.dropna(axis=0, subset=['Embarked'], inplace=True)
df.isna().sum()

PassengerId         0
Survived            0
Pclass              0
Name                0
Sex                 0
Age                 0
SibSp               0
Parch               0
Ticket              0
Fare                0
Embarked            0
Coluna Aleatoria    0
dtype: int64

##Aplicar funções em colunas

In [83]:
#categoria a idade
def calc_faixa_etaria(idade):
  if idade < 13:
    return 'Criança'
  elif idade < 18:
    return 'Adolescente'
  elif idade < 60:
    return 'Adulto'
  else:
    return 'Idoso'

In [85]:
faixa_etarias = df['Age'].apply(calc_faixa_etaria)
faixa_etarias[:5]

1    Adulto
2    Adulto
3    Adulto
4    Adulto
5    Adulto
Name: Age, dtype: object

In [87]:
df.loc[:,'Faixa etaria'] = faixa_etarias
df.head()

PassengerId  Survived  Pclass  ... Embarked Coluna Aleatoria  Faixa etaria
1          2.0       1.0     1.0  ...        C             13.0        Adulto
2          3.0       1.0     3.0  ...        S             28.0        Adulto
3          4.0       1.0     1.0  ...        S              7.0        Adulto
4          5.0       0.0     3.0  ...        S             94.0        Adulto
5          6.0       0.0     3.0  ...        Q             57.0        Adulto

[5 rows x 13 columns]

##Salvar df

In [91]:
df.to_csv('titanic_2.csv', index=False)